transfer-syntax-registry/src/adapters/jpegls.rs

```diff
@@ -93,3 +102,154 @@ impl PixelDataReader for JpegLSAdapter {
+        let frame_size =
+            cols as usize * rows as usize * samples_per_pixel as usize * bytes_per_sample;
+
+        // identify frame data using the frame index
+        let pixeldata_uncompressed = &src
+            .raw_pixel_data()
+            .context(encode_error::MissingAttributeSnafu { name: "Pixel Data" })?
+            .fragments[0];
+
+        let frame_data = pixeldata_uncompressed
+            .get(frame_size * frame as usize..frame_size * (frame as usize + 1))
+            .whatever_context("Frame index out of bounds")?;
+
+        // Encode the data
+        let mut encoder = CharLS::default();
+
+        let frame_info = FrameInfo {
+            width: cols as u32,
+            height: rows as u32,
+            bits_per_sample: bits_stored as i32,
+            component_count: samples_per_pixel as i32,
+        };
+
+        // prefer lossless encoding by default
+        let quality = options.quality.map(|q| q.clamp(0, 100)).unwrap_or(100);
+
+        // calculate the maximum acceptable error range
+        // based on the requested quality and bit depth
+        let near = ((1 << (bits_stored - 4)) * (100 - quality as i32) / 100).min(4096);
+
+        let compressed_data = encoder
+            .encode(frame_info, near, frame_data)
+            .whatever_context("JPEG-LS encoding failed")?;
+
+        dst.extend_from_slice(&compressed_data);
+
+        let mut changes = if near > 0 {
+            let compressed_frame_size = compressed_data.len();
+
+            let compression_ratio = frame_size as f64 / compressed_frame_size as f64;
+            let compression_ratio = format!("{:.6}", compression_ratio);
+
+            // provide attribute changes
+            vec![
+                // lossy image compression
+                AttributeOp::new(Tag(0x0028, 0x2110), AttributeAction::SetStr("01".into())),
+                // lossy image compression ratio
+                AttributeOp::new(
+                    Tag(0x0028, 0x2112),
+                    AttributeAction::PushStr(compression_ratio.into()),
+                ),
+            ]
+        } else {
+            vec![
+                // lossless image compression
+                AttributeOp::new(Tag(0x0028, 0x2110), AttributeAction::SetIfMissing("00".into())),
+            ]
+        };
+
+        let pmi = src.photometric_interpretation();
+
+        if samples_per_pixel == 1 {
+            // set Photometric Interpretation to Monochrome2
+            // if it was neither of the expected monochromes
+            if pmi != Some("MONOCHROME1") && pmi != Some("MONOCHROME2") {
+                changes.push(AttributeOp::new(
+                    Tag(0x0028, 0x0004),
+                    AttributeAction::SetStr("MONOCHROME2".into()),
+                ));
+            }
+        } else if samples_per_pixel == 3 {
+            // set Photometric Interpretation to RGB
+            // if it was not already set to RGB
+            if pmi != Some("RGB") {
```

Comment:
I don't think doing this is safe either, what if the stream is in YBR_FULL, it will change the pi without converting the stream encoding.